### DISCUSSION TITLE
Use uvloop as main HA loop

### DISCUSSION THREAD

**Discussion by rlippmann** [home-assistant/architecture#904] [General] [answered]
The default python asyncio loop is slow.  uvloop is much faster and relatively simple to implement.

I was told uvloop was the default async loop in past versions of HA (circa 2019).  I'm not sure why it was removed, but it seems that the latest version (0.17) of uvloop works.

I propose using uvloop again, unless of course, there are other issues from the past which still haven't been resolved.

[PR link](https://github.com/home-assistant/core/pull/93173) with links to forum post with benchmark results.

**Comment by bdraco** [comment]
We had a lot of problems with executing subprocesses with uvloop.

I think they are fixed by now, but we need to do some archeology to find the old issues to verify they are solved before we could move this forward.

**Comment by bdraco** [comment]
See https://github.com/home-assistant/core/pull/29835 for the history

**Comment by bdraco** [comment]
I tried your PR on my production and everything seems to be working well now.  The subprocess issues do seem to be sovled.

One problem that would need to be overcome is uvloop doesn't support the `<broadcast>` special case https://github.com/python/cpython/blob/5e9f471e7db30893fb3f42681f17fdcdb70069ee/Modules/socketmodule.c#L1169

which means the following would break which makes this a non-starter at the moment

```
./lib/python3.10/site-packages/jellyfin_apiclient_python/connection_manager.py:        MULTI_GROUP = ("<broadcast>", 7359)
./lib/python3.10/site-packages/jaraco/net/wake.py:    s.sendto(magic_packet, ('<broadcast>', 9))
./lib/python3.10/site-packages/elkm1_lib/discovery.py:    BROADCAST_ADDRESS = "<broadcast>"
./lib/python3.10/site-packages/pycomfoconnect/bridge.py:            udpsocket.sendto(b"\x0a\x00", ('<broadcast>', Bridge.PORT))
./lib/python3.10/site-packages/miio/miioprotocol.py:            addr = "<broadcast>"
./lib/python3.10/site-packages/discovery30303/__init__.py:    BROADCAST_ADDRESS = "<broadcast>"
./lib/python3.10/site-packages/nmb/NetBIOSProtocol.py:        # We don't use the transport.write method directly as it keeps raising DeprecationWarning for ip='<broadcast>'
./lib/python3.10/site-packages/nmb/NetBIOSProtocol.py:            ip = '<broadcast>'
./lib/python3.10/site-packages/nmb/NetBIOS.py:            ip = '<broadcast>'
./lib/python3.10/site-packages/flux_led/scanner.py:    BROADCAST_ADDRESS = "<broadcast>"
./lib/python3.10/site-packages/aiosenseme/discovery.py:            self.transport.sendto(data, ("<broadcast>", PORT))
./lib/python3.10/site-packages/pyric/pyw.py:     set nic's ip4 netmask (ifconfig <card.dev> broadcast <broadcast>
./lib/python3.10/site-packages/unifi_discovery/__init__.py:            address = "<broadcast>"
./lib/python3.10/site-packages/roonapi/discovery.py:            sock.sendto(msg, ("<broadcast>", SOOD_PORT))
./lib/python3.10/site-packages/pybravia/client.py:            sock.sendto(packet, ("<broadcast>", 9))
./lib/python3.10/site-packages/roombapy/discovery.py:    udp_address = "<broadcast>"
```

**Comment by bdraco** [comment, marked as answer]
The performance improvements for sockets that read frequently (ie bluetooth, zeroconf, etc) is quite stunning and this is solving a problem we actually do have so it would be nice if we could get this working but the `<broadcast>` issue is definitely a blocker here.

**Comment by rlippmann** [comment]
Might this be part of the problem?

[uvloop is broken for abstract sockets](https://github.com/MagicStack/uvloop/issues/528)

**Reply by bdraco** [reply]
Thats a different issue, but will likely also been an issue for dbus

**Comment by rlippmann** [comment]
I'm already subclassing the event loop in the pr, maybe override socket calls for broadcast?

Or log an issue against uvloop?

**Reply by bdraco** [reply]
The discussion on how to solve the problems should take place in the PR. Once you have a solution to solve both of these issues than we can come back here and do more testing to see if there are still blocking issues

**Comment by Jc2k** [comment]
Does this change anything?

https://github.com/python/cpython/issues/91166

(https://github.com/python/cpython/issues/91166#issuecomment-1363934346)

**Reply by bdraco** [reply]
That will help quite a bit

**Comment by rlippmann** [comment]
3.11 also has an async sendto

**Comment by rlippmann** [comment]
So, there is a [PR to fix `'<broadcast>'`](https://github.com/MagicStack/uvloop/pull/592), but it's been outstanding since September.

It's a pretty simple change (most of the PR is actually just updating the tests).

Since the access to the private timer variable in asyncio has been fixed, perhaps fork uvloop into another pypi package until this gets merged?

**Reply by bdraco** [reply]
Given the data leakage problem has gone unaddressed it seems like uvloop is barely maintained these days https://github.com/MagicStack/uvloop/pull/646.  While there is still a nice speed up, its getting less and less relevant as Python 3.13 has even more improvements in event loop performance.

**Reply by rlippmann** [reply]
As an update, the PR was finally merged today.  But the other issue is still open.
